compiler/lib/parse_bytecode.ml

```diff
@@ -334,19 +334,41 @@ end = struct
   let rec scan debug blocks code pc len =
     if pc < len
     then
-      match (get_instr_exn code pc).kind with
+      let instr = get_instr_exn code pc in
+      match instr.kind with
       | KNullary -> scan debug blocks code (pc + 1) len
       | KUnary -> scan debug blocks code (pc + 2) len
       | KBinary -> scan debug blocks code (pc + 3) len
       | KNullaryCall ->
           let blocks =
             if Debug.mem debug (pc + 1) then Addr.Set.add pc blocks else blocks
           in
+          let blocks =
+            (* The code transformation used to deal with effect
+               handlers expects that function applications and effect
+               primitives are at the very end of a block. *)
```

Comment:
We could have a pass right before the effect  one to put the program in the required form, That way we have less invariant to preserve and we don't have to disable optimisations breaking this invariants.
Also, less blocks would probably increase compilation speed (see timing in https://github.com/ocsigen/js_of_ocaml/issues/1300#issuecomment-1276796143)

Reply:
I've done that. There are some speed improvements, indeed (about 10% faster on the Ocsigen Start demo).

Reply:
> about 10% faster on the Ocsigen Start demo.

Is this will or without debug events ? (e.g. --pretty)